src/wp-includes/class-wp-token-map.php

```diff
@@ -0,0 +1,810 @@
+	 * byte. This makes for quick lookup of where in the group string the key
+	 * is found, and then a simple division converts that offset into the index
+	 * in the `$large_words` array where the group string is to be found.
+	 *
+	 * This lookup data structure is designed to optimize cache locality and
+	 * minimize indirect memory reads when matching strings in the set.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var array
+	 */
+	private $large_words = array();
+
+	/**
+	 * Stores the group keys for sequential string lookup.
+	 *
+	 * The offset into this string where the group key appears corresponds with the index
+	 * into the group array where the rest of the group string appears. This is an optimization
+	 * to improve cache locality while searching and minimize indirect memory accesses.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var string
+	 */
+	private $groups = '';
+
+	/**
+	 * Stores an optimized row of small words, where every entry is
+	 * `$this->key_size + 1` bytes long and zero-extended.
+	 *
+	 * This packing allows for direct lookup of a short word followed
+	 * by the null byte, if extended to `$this->key_size + 1`.
+	 *
+	 * Example:
+	 *
+	 *     // Stores array( 'GT', 'LT', 'gt', 'lt' ).
+	 *     "GT\x00LT\x00gt\x00lt\x00"
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var string
+	 */
+	private $small_words = '';
+
+	/**
+	 * Replacements for the small words, in the same order they appear.
+	 *
+	 * With the position of a small word it's possible to index the translation
+	 * directly, as its position in the `$small_words` string corresponds to
+	 * the index of the replacement in the `$small_mapping` array.
+	 *
+	 * Example:
+	 *
+	 *     array( '>', '<', '>', '<' )
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var string[]
+	 */
+	private $small_mappings = array();
+
+	/**
+	 * Create a token map using an associative array of key/value pairs as the input.
+	 *
+	 * Example:
+	 *
+	 *     $smilies = WP_Token_Map::from_array( array(
+	 *          '8O' => '😯',
+	 *          ':(' => '🙁',
+	 *          ':)' => '🙂',
+	 *          ':?' => '😕',
+	 *       ) );
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param array $mappings   The keys transform into the values, both are strings.
+	 * @param int   $key_length Determines the group key length. Leave at the default value
+	 *                          of 2 unless there's an empirical reason to change it.
+	 *
+	 * @return WP_Token_Map|null Token map, unless unable to create it.
+	 */
+	public static function from_array( $mappings, $key_length = 2 ) {
+		$map             = new WP_Token_Map();
+		$map->key_length = $key_length;
+
+		// Start by grouping words.
+
+		$groups = array();
+		$shorts = array();
+		foreach ( $mappings as $word => $mapping ) {
+			if (
+				self::MAX_LENGTH <= strlen( $word ) ||
+				self::MAX_LENGTH <= strlen( $mapping )
+			) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Token Map tokens and substitutions must all be shorter than 256 bytes.' ),
```

Comment:
Minor: Rather than hardcoding 256, I think this should use `self::MAX_LENGTH` on the chance it changes in the future. 
```suggestion
					/* translators: maximum byte length */ 
					sprintf( __( 'Token Map tokens and substitutions must all be shorter than %i bytes.' ), self::MAX_LENGTH ),
```

Reply:
good catch! incorporated in 0e7ab4ca52d97da25a20e0e493bc33308f56cf1d